### DISCUSSION TITLE
Comparison of ScopeProvider logic in Xtext and Langium

### DISCUSSION THREAD

**Discussion by movalz** [eclipse-langium/langium#1569] [Q&A] [answered]
I come from Xtext (we have some big DSLs defined in Xtext) and investigate switching to Langium. I am not yet an expert in TypeScript, Node.js and VSCode.

One important point is "ScopeProvider" logic. As far as I have understood, the default logic for ScopeProvider differs in following points:
- in Xtext the generated MyGrammarDslScopeProvider can be filled with methods with signature `public IScope scope_TargetType(ContextType ctx, EReference reference)` or `public IScope scope_ReferenceContainingType_ReferenceName(ContextType ctx, EReference reference)` which are automatically called from `AbstractDeclarativeScopeProvider.polymorphicFindScopeForReferenceName()` and `AbstractDeclarativeScopeProvider.polymorphicFindScopeForClassName()`, walking the eContainer path from the current AST node to the root node. I can't find anything similar in Langium. Instead I have to implement a ScopeProvider (extending DefaultScopeProvider) without any implementation support like AbstractDeclarativeScopeProvider.polymorphicFindScopeFor... Is this correct? Is it planned to implement a similar approach in the future?
- global scope ("index") - Xtext's DefaultGlobalScopeProvider loops over all "visible containers". These are all models of my language which are contained in the classpath. In Langium (AFAIK) all models are processed which belong to the "workspaceFolders" of VSCode (and sub folders). How can I include models which are contained in another node_module I depend on (having a base node.js project which contains some base models)?

The logic for determination of the local scope in Xtext and Langium seems to be identical (as described in [DefaultScopeComputation.ts](https://github.com/eclipse-langium/langium/blob/721b882c8f2e9eecfcc0ba445876ef1c4dd1ed21/packages/langium/src/references/scope-computation.ts#L55-L65)), am I right? (thank you for inline doc, that's missing in Xtext)

**Comment by movalz** [comment]
No response? It would be nice to get an answer for my questions.

**Comment by msujew** [comment, marked as answer]
Hey @movalz, sorry for the late response, it seems like the notification for this fell through the cracks.

 > Is this correct? Is it planned to implement a similar approach in the future?

I believe the `polymorphicFindScopeForReferenceName` and `polymorphicFindScopeForClassName` methods were always pretty hacky. Since we aim to keep the framework as lean as possible, we went with the simplest design for the `ScopeProvider` available. Note that you can easily reimplement that behavior yourself in TypeScript:

```ts
getScope(context: ReferenceInfo): Scope {
  const referenceName = `scope_${context.container.$type}_${context.property}`;
  if (referenceName in this) {
    return this[referenceName](context);
  } else {
    return super.getScope(context);
  }
}
```

> How can I include models which are contained in another node_module I depend on (having a base node.js project which contains some base models)?

You can add arbitrary directories to the workspace, see also the [WorkspaceManager implementation](https://github.com/eclipse-langium/langium/blob/fad57e91a63a57125e372d811eca1c1702c858a2/packages/langium/src/workspace/workspace-manager.ts#L95-L101). You can also add documents individually by using the `LangiumDocuments#getOrCreateDocument` and `addDocument` methods. Loading the VS Code workspace folders is just the default implementation.

> The logic for determination of the local scope in Xtext and Langium seems to be identical (as described in [DefaultScopeComputation.ts](https://github.com/eclipse-langium/langium/blob/721b882c8f2e9eecfcc0ba445876ef1c4dd1ed21/packages/langium/src/references/scope-computation.ts#L55-L65)), am I right?

Yes, even though the implementation differs a bit (Langium uses a less resource intensive method to determine local scopes) the result should generally be the same.
